File: Sources/App/DTOs/AlbumDTO.swift
//
//  AlbumDTO.swift
//  FinchServer
//
//  Created by Stefan Klein Nulent on 25/01/2025.
//

import Vapor

struct AlbumDTO: Content {
    
    // MARK: - Types
    
    enum CodingKeys: String, CodingKey {
        case id
        case title
        case releaseGroupTitle = "release_group_title"
        case artist
        case artistSortKey = "artist_sort_key"
        case type
        case genre
        case style
        case year
        case discCount = "disc_count"
        case discogsAlbumId = "discogs_album_id"
        case discogsArtistId = "discogs_artist_id"
        case discogsLabelId = "discogs_label_id"
        case label
        case barcode
        case asin
        case catalogNumber = "catalog_number"
        case country
        case isArtworkAvailable = "is_artwork_available"
        case artworkPath = "artwork_path"
        case addedAt = "added_at"
        case media
        case dataSource = "data_source"
        case items
    }
    
    
    
    // MARK: - Properties
    
    let id: Int?
    let title: String
    let releaseGroupTitle: String
    let artist: String
    let artistSortKey: String
    let type: String
    let genre: String?
    let style: String?
    let year: Int
    let discCount: Int
    let discogsAlbumId: Int?
    let discogsArtistId: Int?
    let discogsLabelId: Int?
    let label: String?
    let barcode: String?
    let asin: String?
    let catalogNumber: String?
    let country: String?
    let isArtworkAvailable: Bool
    let artworkPath: String?
    
    let addedAt: Date
    let media: String?
    let dataSource: String?
    
    let items: [ItemDTO]?
    
    
    
    // MARK: - Construction
    
    init(_ album: Album, items: [Item]? = nil) {
        id = album.id
        title = album.title
        releaseGroupTitle = album.releaseGroupTitle
        artist = album.albumArtist
        artistSortKey = album.albumArtistSortKey
        type = album.albumType
        genre = album.genre.nilIfEmpty
        style = album.style.nilIfEmpty
        year = album.year
        discCount = album.discCount
        discogsAlbumId = album.discogsAlbumId != 0 ? album.discogsAlbumId : nil
        discogsArtistId = album.discogsArtistId != 0 ? album.discogsArtistId : nil
        discogsLabelId = album.discogsLabelId != 0 ? album.discogsLabelId : nil
        label = album.label.nilIfEmpty
        barcode = album.barcode.nilIfEmpty
        asin = album.asin.nilIfEmpty
        catalogNumber = album.catalogNumber.nilIfEmpty
        country = album.country
        isArtworkAvailable = album.isArtworkAvailable
        artworkPath = album.artworkPath
        addedAt = album.addedAt
        media = album.attributes.first{ $0.key == "media" }?.value
        dataSource = album.attributes.first{ $0.key == "data_source" }?.value
        
        if let items {
            self.items = items.map{ ItemDTO($0) }
        } else {
            self.items = nil
        }
    }
}

File: FinchServer/Controllers/DTO/AlbumDto.cs
using FinchServer.Beets;

namespace FinchServer.Controllers.DTO;

public class AlbumDto(Album album) {
    
    // - Properties
    
    public int Id { get; set; } = album.Id;
    public string Title { get; set; } = album.Title;
    public string Type { get; set; } = album.AlbumType;
    public string Artist { get; set; } = album.AlbumArtist;
    public string? Genre { get; set; } = album.Genre;
    public string? Style { get; set; } = album.Style;
    public string? Country { get; set; } = album.Country;
    public int Year { get; set; } = album.Year;
    public int DiscCount { get; set; } = album.DiscCount;
    public int? DiscogsAlbumId { get; set; } = (album.DiscogsAlbumId ?? 0) > 0 ? album.DiscogsAlbumId : null;
    public int? DiscogsArtistId { get; set; } = (album.DiscogsArtistId ?? 0) > 0 ? album.DiscogsArtistId : null;
    public int? DiscogsLabelId { get; set; } = (album.DiscogsLabelId ?? 0) > 0 ? album.DiscogsLabelId : null;
    public DateTimeOffset AddedAt { get; set; } = DateTimeOffset.FromUnixTimeSeconds((long)album.Added);
    public string? Label { get; set; } = album.Label;
    public string? Barcode { get; set; } = !string.IsNullOrEmpty(album.Barcode) ? album.Barcode : null;
    public string? Asin { get; set; } = !string.IsNullOrEmpty(album.Asin) ? album.Asin : null;
    public string? CatalogNumber { get; set; } = !string.IsNullOrEmpty(album.CatalogNumber) ? album.CatalogNumber : null;
    public string? ArtworkPath { get; set; } = album.ArtworkPath;
    public bool IsArtworkAvailable { get; set; } = album.IsArtworkAvailable;
    public ItemDto[]? Items { get; set; } = album.Items?.Select(i => new ItemDto(i)).ToArray();
}